### ISSUE TITLE
Link libdispatch on platforms where it isn't present in libSystem

### ISSUE THREAD

**Issue by bbqsrc** [SSheldon/rust-dispatch#3]
Apple recently open sourced Swift, and brought with it an up-to-date [libdispatch](https://github.com/apple/swift-corelibs-libdispatch) that they intend to make work correctly on Linux.

I'm wondering whether it would be worthwhile to rebase onto that target?

**Comment by SSheldon** [comment]
Hey @bbqsrc, this crate just links against whatever libdispatch is available, it doesn't compile its own libdispatch or anything like that. As far as I understand, Swift's libdispatch will maintain an identical API to Apple's, so this crate should work just fine regardless of which implementation you use.

Are there any differences in the public API for Swift's libdispatch? Feel free to open issues if you come across some!

**Comment by bbqsrc** [comment]
Sure, but what made me ask this was this [line](https://github.com/SSheldon/rust-dispatch/blob/master/src/ffi.rs#L27) which links to `libSystem`, which only exists on OS X. Unless I misunderstand how FFI works with Rust (and I probably do), this means it'll always try `-lSystem` and nothing else.

The library is known by `libdispatch` on other platforms, and requires a bunch of other linked libraries like the block library and `kqueue` emulation.

**Comment by SSheldon** [comment]
Ah I understand @bbqsrc, that's a good point. I'll reopen this issue but probably will need help from someone with a Linux system to solve this :)

**Comment by bbqsrc** [comment]
That's a better title. :+1:

I'm watching that repo for Linux support, so the second it's there, I'll be fiddling with rust-dispatch until I get it to link then I'll make a pull request. :smile:

**Comment by SSheldon** [comment]
Great, thanks @bbqsrc!

**Comment by luxvio** [comment]
Aside from OS X, FreeBSD also has support for libdispatch and it's deps are practically all in the base system/kernel ... but on the other side Rust support is kinda salty (sometimes works, sometimes bang in the stack ! ... really sad, because I think Rust+FreeBSD could be an awesome combination for server development)

Probably the best idea from the very start would be to link libSystem on apple's systems and libdispatch on everything else. But ATM we should pretty much wait they settle on the interfaces while they ready libdispatch for more cross-platform duty (afaik, a subset of the library will cross-platform and other functions won't)

**Comment by luxvio** [comment]
Ok, recently gave another shot at compiling libdispatch (in the middle of playing with gnustep svn) on Linux and everything seems ok. I downloaded `rust-dispatch` and changed the link line to something like this:

```rust
#[cfg_attr(not(target_os = "macos"), link(name = "dispatch", kind = "dylib") )]
#[cfg_attr(target_os = "macos", link(name = "System", kind = "dylib") )]
```

Bam ! everything seems to work and test passes ... imported on another project using path to import the crate and all fine in my fiddlings, apparently this **does** use a subset that is ok with ported dispatch.

Regards.

P.S → On FreeBSD, it's also `-ldispatch`

**Comment by bbqsrc** [comment]
:+1:

**Comment by SSheldon** [comment]
Thanks @luzvioleta and @bbqsrc! I finally got around to setting up travis ci for this repo, so I was able to add a linux configuration to verify this. Merged the fix in 202898d.
